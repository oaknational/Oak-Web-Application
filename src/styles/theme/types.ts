import { ThemedStyledProps, DefaultTheme } from "styled-components";
import { z } from "zod";

import { BadgeConfig } from "../../components/Badge";
import { CheckboxConfig } from "../../components/Checkbox";
import { SelectListBoxConfig } from "../../components/DropdownSelect/ListBox";
import { HeaderConfig } from "../../components/FixedHeader/FixedHeader";
import { MenuConfig } from "../../components/Menu/Menu";
import { ToggleStyleConfig } from "../../components/Toggle/Toggle";
import { VideoStyleConfig } from "../../components/VideoPlayer/VideoPlayer";

/**
 * Adds a finite list of pixel values which we're allowed to use throughout the
 * app's styles.
 * Keeping this list finite and explicit means that that (1) consistency is
 * encouraged and (2) if we want to migrate our styles to a static css solution
 * e.g. Vanilla Extract, then we can (otherwise we rely on runtime css to
 * generate classes for adhoc pixel values).
 * @todo use negative number type
 * @see https://stackoverflow.com/questions/21224922/is-there-a-way-to-represent-a-non-negative-integer-in-typescript-so-that-the-com
 */
export type PixelSpacing =
  | 0
  | 1
  | 3
  | 4
  | 6
  | 8
  | 10
  | 12
  | 16
  | 18
  | 20
  | 24
  | 28
  | 30
  | 32
  | 36
  | 40
  | 44
  | 48
  | 50
  | 56
  | 64
  | 72
  | 80
  | 92
  | 96
  | 110
  | 120
  | 130
  | 140
  | 150
  | 160
  | 166
  | 172
  | 200
  | 220
  | 240
  | 270
  | 300
  | 320
  | 360
  | 380
  | 400
  | 420
  | 450
  | 480
  | 580
  | 600
  | 640
  | 720
  | 740
  | 812
  | 840
  | 900
  | 960
  | 1280
  | 1600;
export type NullablePixelSpacing = PixelSpacing | null;
export type NegativePixelSpacing =
  | -36
  | -32
  | -28
  | -24
  | -20
  | -16
  | -12
  | -8
  | -4;
export type PercentSpacing =
  | "100%"
  | "99%"
  | "95%"
  | "90%"
  | "85%"
  | "80%"
  | "70%"
  | "60%"
  | "55%"
  | "50%"
  | "40%"
  | "30%"
  | "33%"
  | "25%"
  | "20%"
  | "15%"
  | "10%"
  | "5%"
  // vw units sometimes needed to fix edge cages
  | "28vw";
export type Unset = "unset";

/**
 * @todo parse theme with zod
 */
const OakColorNameZod = z.union([
  z.literal("white"),
  z.literal("black"),
  z.literal("transparent"),
  z.literal("inherit"),
  z.literal("grey1"),
  z.literal("grey2"),
  z.literal("grey3"),
  z.literal("grey4"),
  z.literal("grey5"),
  z.literal("grey6"),
  z.literal("grey7"),
  z.literal("grey8"),
  z.literal("grey9"),
  z.literal("grey10"),
  z.literal("oakGrey1"),
  z.literal("oakGrey2"),
  z.literal("oakGrey3"),
  z.literal("oakGrey4"),
  z.literal("oakGrey5"),
  z.literal("oakGrey6"),
  z.literal("pastelTurquoise"),
  z.literal("warning"),
  z.literal("failure"),
  z.literal("success"),
  z.literal("pupilsHighlight"),
  z.literal("pupilsAccentPink"),
  z.literal("pupilsGreen"),
  z.literal("pupilsLightGreen"),
  z.literal("pupilsLimeGreen"),
  z.literal("pupilsPink"),
  z.literal("teachersGreen"),
  z.literal("teachersYellow"),
  z.literal("teachersPastelYellow"),
  z.literal("teachersPastelBlue"),
  z.literal("teachersRed"),
  z.literal("teachersPurple"),
  z.literal("teachersLilac"),
  z.literal("teachersHighlight"),
  z.literal("twilight"),
  z.literal("videoBlue"),
  z.literal("hyperlink"),
]);

export type OakColorName = z.infer<typeof OakColorNameZod>;
export type OakFontName = "body" | "ui" | "heading" | "headingLight";
/**
 * ColorValue could be hex, rgb, rgba, hsla, e.g. "#414243"
 */
type ColorValue = string;
/**
 * FontValue should be a font name with fallback(s), e.g. "Lexend, sans-serif"
 */
type FontValue = string;

type ButtonConfig = {
  disabled: {
    background: OakColorName;
    text: OakColorName;
  };
  primary: {
    background: OakColorName;
    text: OakColorName;
  };
  secondary: {
    background: OakColorName;
    text: OakColorName;
  };
  tertiary: {
    background: OakColorName;
    text: OakColorName;
  };
  minimal: {
    background: OakColorName;
    text: OakColorName;
  };
};
type InputConfig = {
  height: string;
  borderRadius: string;
  borderWidth: string;
  fontSize: string;
  states: {
    default: InputStateConfig;
    active: InputStateConfig;
    valid: InputStateConfig;
    invalid: InputStateConfig;
    disabled: InputStateConfig;
  };
};

type InputStateConfig = {
  text: OakColorName;
  placeholder: OakColorName;
  icon: OakColorName;
  border: OakColorName;
  background: OakColorName;
};

export type OakTheme = {
  header: HeaderConfig;
  name: string;
  colors: Record<OakColorName, ColorValue>;
  contrastColors: Record<OakColorName, OakColorName>;
  buttonIconBackgroundColors: Partial<Record<OakColorName, OakColorName>>;
  buttonDropShadows: Partial<Record<OakColorName, string>>;
  buttonFocusUnderlineColors: Partial<Record<OakColorName, OakColorName>>;
  fonts: Record<OakFontName, FontValue>;
  input: InputConfig;
  button: ButtonConfig;
  badge: BadgeConfig;
  checkbox: CheckboxConfig;
  selectListBox: SelectListBoxConfig;
  toggle: ToggleStyleConfig;
  video: VideoStyleConfig;
  menu: MenuConfig;
};

export type PropsWithTheme<Props = unknown> = ThemedStyledProps<
  Props,
  DefaultTheme
>;
