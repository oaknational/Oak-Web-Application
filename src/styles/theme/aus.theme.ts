import { OakTheme, default as defaultTheme } from ".";

const ausTheme: OakTheme = {
  // Extends default theme
  ...defaultTheme,
  name: "aus",
  button: {
    ...defaultTheme.button,
    // Simply altering primary button config as an example
    primary: {
      background: "teachersLilac",
      text: "white",
    },
  },
  toggle: {
    on: {
      labelColor: "grey10",
      background: "white",
      switchColor: "grey8",
    },
    off: {
      labelColor: "grey8",
      background: "white",
      switchColor: "grey8",
    },
    disabled: {
      labelColor: "grey3",
      background: "white",
      switchColor: "grey6",
    },
  },
};

export default ausTheme;
