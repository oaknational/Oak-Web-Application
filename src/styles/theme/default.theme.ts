import { OakTheme } from "./types";

const theme: OakTheme = {
  name: "default",
  colors: {
    white: "#fff",
    black: "#000",
    transparent: "transparent",
    inherit: "inherit",
    grey1: "#f2f2f2",
    grey2: "#e6e6e6",
    grey3: "#ccc",
    grey4: "#b3b3b3",
    grey5: "#999",
    grey6: "#808080",
    grey7: "#666",
    grey8: "#4d4d4d",
    grey9: "#333",
    grey10: "#1a1a1a",
    inYourFace: "#ef476f",
    calmAndWarm: "#ffd166",
    niceAndSharp: "#06d6a0",
    deeperWins: "#073b4c",
    limeade: "#32be00",
    madangGreen: "#bef2bd",
    mustard: "#ffe555",
    peranoBlue: "#a0b6f2",
    powderBlue: "#b0e2de",
    melaniePink: "#deb7d5",
    error: "red",
  },
  contrastColors: {
    white: "black",
    black: "white",
    transparent: "inherit",
    inherit: "inherit",
    grey1: "black",
    grey2: "black",
    grey3: "black",
    grey4: "black",
    grey5: "black",
    grey6: "white",
    grey7: "white",
    grey8: "white",
    grey9: "white",
    grey10: "white",
    inYourFace: "white",
    calmAndWarm: "grey9",
    niceAndSharp: "grey8",
    deeperWins: "white",
    limeade: "black",
    madangGreen: "black",
    mustard: "black",
    peranoBlue: "black",
    powderBlue: "black",
    melaniePink: "black",
    error: "white",
  },
  fonts: {
    // Paragraphs etc.
    body: "ABeeZee, sans-serif",
    // Buttons etc.
    ui: "Lexend, sans-serif",
    // Headings etc.
    heading: "Lexend, sans-serif",
  },
  input: {
    height: "40px",
    borderRadius: "8px",
    borderWidth: "1px",
    states: {
      default: {
        text: "black",
        placeholder: "grey6",
        icon: "grey6",
        border: "grey6",
      },
      active: {
        text: "black",
        placeholder: "grey6",
        icon: "grey6",
        border: "grey8",
      },
      valid: {
        text: "black",
        placeholder: "grey6",
        icon: "grey8",
        border: "grey8",
      },
      invalid: {
        text: "black",
        placeholder: "grey6",
        icon: "grey6",
        border: "grey8",
      },
      disabled: {
        text: "black",
        placeholder: "grey6",
        icon: "grey6",
        border: "grey8",
      },
    },
  },
  bigInput: {
    height: "40px",
    borderRadius: "20px",
    borderWidth: "0",
    states: {
      default: {
        text: "black",
        placeholder: "grey6",
        icon: "grey6",
        border: "grey8",
      },
      active: {
        text: "black",
        placeholder: "grey6",
        icon: "grey6",
        border: "grey8",
      },
      valid: {
        text: "black",
        placeholder: "grey6",
        icon: "grey6",
        border: "grey8",
      },
      invalid: {
        text: "black",
        placeholder: "grey6",
        icon: "grey6",
        border: "grey8",
      },
      disabled: {
        text: "black",
        placeholder: "grey6",
        icon: "grey6",
        border: "grey8",
      },
    },
  },
  button: {
    disabled: {
      background: "grey6",
      text: "white",
    },
    primary: {
      background: "black",
      text: "white",
    },
    secondary: {
      background: "grey9",
      text: "white",
    },
    tertiary: {
      background: "grey3",
      text: "black",
    },
  },
  badge: {
    size: "54px",
    circleSize: "48px",
    fontSize: "16px",
    iconSize: 20,
    starColor: "grey6",
    circleColor: "grey8",
    textColor: "white",
  },
  lessonControl: {
    default: {
      background: "white",
      border: "2px solid",
      borderColor: "transparent",
    },
    current: {
      background: "grey3",
      border: "2px dashed",
      borderColor: "niceAndSharp",
    },
    complete: {
      background: "grey7",
      border: "2px solid",
      borderColor: "transparent",
    },
  },
  checkbox: {
    default: {
      color: "black",
    },
    disabled: {
      color: "grey4",
    },
  },
  toggle: {
    on: {
      labelColor: "grey10",
      background: "white",
      switchColor: "grey8",
    },
    off: {
      labelColor: "grey8",
      background: "white",
      switchColor: "grey8",
    },
    disabled: {
      labelColor: "grey3",
      background: "white",
      switchColor: "grey6",
    },
  },
};

export default theme;
