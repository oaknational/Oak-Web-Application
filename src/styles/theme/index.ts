export { default } from "./default.theme";
export type {
  OakTheme,
  OakColorName,
  OakFontName,
  PropsWithTheme,
} from "./types";
