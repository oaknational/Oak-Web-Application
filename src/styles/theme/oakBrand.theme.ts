import { OakTheme, default as defaultTheme } from ".";

const oakBrandTheme: OakTheme = {
  // Extends default theme
  ...defaultTheme,
  name: "Oak Brand",
  buttons: {
    ...defaultTheme.buttons,
    primary: {
      background: "oakGreen",
      text: "white",
    },
  },
};

export default oakBrandTheme;
