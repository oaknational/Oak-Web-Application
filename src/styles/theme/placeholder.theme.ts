import { OakTheme, default as defaultTheme } from ".";

const placeholderTheme: OakTheme = {
  // Extends default theme
  ...defaultTheme,
  name: "placeholder",
  button: {
    ...defaultTheme.button,
    // Simply altering primary button config as an example
    primary: {
      background: "limeade",
      text: "white",
    },
  },
};

export default placeholderTheme;
