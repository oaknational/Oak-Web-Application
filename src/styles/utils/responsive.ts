import {
  css,
  DefaultTheme,
  Interpolation,
  ThemedStyledProps,
} from "styled-components";

import truthy from "../../utils/truthy";
import { PropsWithTheme } from "../theme";

const breakpointsByName = {
  small: 750,
  large: 1280,
};
export const breakpoints = Object.values(breakpointsByName).sort((a, b) =>
  a > b ? 1 : -1
);
export type BreakpointName = keyof typeof breakpointsByName;
export const getBreakpoint = (
  breakpointName: keyof typeof breakpointsByName
) => {
  return breakpointsByName[breakpointName];
};
export type Device = "mobile" | "tablet" | "desktop";
const mediaQueries: Record<Device, string> = {
  mobile: `(max-width: ${getBreakpoint("small") - 1}px)`,
  tablet: `(min-width: ${getBreakpoint("small")}px and max-width: ${
    getBreakpoint("large") - 1
  }px)`,
  desktop: `(min-width: ${getBreakpoint("large")}px)`,
};
export const getMediaQuery = (device: Device) => {
  return mediaQueries[device];
};
export type ResponsiveValues<Value> = (Value | null) | (Value | null)[];

const responsive =
  <Props, T extends string | number | undefined | null>(
    attr: string,
    getValues: (props: Props) => ResponsiveValues<T | undefined | null>,
    parse:
      | ((unparsed: T | undefined | null) => string | number | undefined | null)
      | ((
          unparsed: T | undefined | null
        ) => (props: PropsWithTheme) => string | number | undefined | null) = (
      x
    ) => x
  ) =>
  (props: Props): Interpolation<ThemedStyledProps<Props, DefaultTheme>> => {
    const attrCss = (value: T | undefined | null) =>
      typeof value === "undefined"
        ? undefined
        : css`
            ${attr}: ${parse(value)};
          `;
    const values = getValues(props);
    if (typeof values === "undefined") {
      return undefined;
    }
    if (!Array.isArray(values)) {
      return css`
        ${attrCss(values)}
      `;
    }
    if (values.length === 0) {
      return [];
    }

    return [
      css`
        ${attrCss(values[0])}
      `,
      ...breakpoints
        .slice(0, values.length)
        .map((breakpoint, i) => {
          const value = values[i + 1]; // Values are shifted relative to breakpoints

          if (value === undefined) {
            return undefined;
          }

          return css`
            @media (min-width: ${breakpoint}px) {
              ${css`
                ${attrCss(value)}
              `}
            }
          `;
        })
        .filter(truthy),
    ];
  };

export default responsive;
