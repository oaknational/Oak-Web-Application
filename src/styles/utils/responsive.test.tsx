import styled, {
  css,
  DefaultTheme,
  FlattenSimpleInterpolation,
  Interpolation,
  ThemedStyledProps,
} from "styled-components";
import { render } from "@testing-library/react";

import { OakColorName } from "../theme";
import renderWithProviders from "../../__tests__/__helpers__/renderWithProviders";

import responsive from "./responsive";

type TestProps = {
  [k: string]: OakColorName | string | string[] | number | number[];
};
/**
 *
 * @description outputs from styled-components css function can vary in value
 * whilst still representing the same css.
 * For this reason we need to "stringify" (and minify) the css that we get from responsive
 * in order to compare it with the expected css values.
 */
const stringify = (
  cssArray:
    | FlattenSimpleInterpolation
    | Interpolation<ThemedStyledProps<TestProps, DefaultTheme>>
) =>
  (Array.isArray(cssArray) ? cssArray : [cssArray])
    ?.flatMap((str: unknown) => (typeof str === "string" ? str.trim() : str))
    .flat()
    .join("")
    .replace(/([^0-9a-zA-Z.#])\s+/g, "$1")
    .replace(/\s([^0-9a-zA-Z.#]+)/g, "$1")
    .replace(/;}/g, "}")
    .replace(/\/\*.*?\*\//g, "")
    .trim();

const pxOrUndefined = (value: number | unknown) =>
  typeof value === "number" ? `${value}px` : undefined;

describe("responsive", () => {
  it("should correctly handle a single value", async () => {
    const props = {
      pl: 12,
    };
    const styles = responsive(
      "padding-left",
      (props: TestProps) => props.pl,
      pxOrUndefined
    )(props);
    const StyledComponent = styled.div`
      ${styles}
    `;
    const { getByTestId } = render(
      <StyledComponent data-testid="test" pl={12} />
    );

    expect(getByTestId("test")).toHaveStyle("padding-left: 12px");
  });
  it("should correctly handle array of values", async () => {
    const props = {
      pl: [0, 12],
    };

    const actual = responsive(
      "padding-left",
      (props: TestProps) => props.pl,
      pxOrUndefined
    )(props);
    const expected = css`
      padding-left: 0px;

      @media (min-width: 600px) {
        padding-left: 12px;
      }
    `;

    expect(stringify(actual)).toEqual(stringify(expected));
  });
  it("should handle the case where the last value is a zero", async () => {
    const props = {
      pl: [36, 12, 6, 0],
    };

    const actual = responsive(
      "padding-left",
      (props: TestProps) => props.pl,
      pxOrUndefined
    )(props);
    const expected = css`
      padding-left: 36px;

      @media (min-width: 600px) {
        padding-left: 12px;
      }

      @media (min-width: 900px) {
        padding-left: 6px;
      }

      @media (min-width: 1200px) {
        padding-left: 0px;
      }
    `;
    expect(stringify(actual)).toEqual(stringify(expected));
  });
  it("should default parse to be identity fn", () => {
    const props = {
      pl: "0.5em",
    };

    const actual = responsive(
      "padding-left",
      (props: TestProps) => props.pl
    )(props);
    const expected = css`
      padding-left: 0.5em;
    `;
    expect(stringify(actual)).toEqual(stringify(expected));
  });
  test("should handle when parse fn gets from theme", async () => {
    const StyledComponent = styled.div`
      ${responsive(
        "color",
        (props) => props.color,
        (colorName) => (props) => props.theme.colors[colorName as OakColorName]
      )}
    `;
    const { getByTestId } = renderWithProviders(
      <StyledComponent data-testid="test" color="calmAndWarm" />
    );

    expect(getByTestId("test")).toHaveStyle("color: #ffd166");
  });
  test.each([
    ["pl", "padding-left", "1em", "padding-left: 1em;"],
    ["pr", "padding-right", "1em", "padding-right: 1em;"],
    ["pt", "padding-top", "1em", "padding-top: 1em;"],
    ["pb", "padding-bottom", "1em", "padding-bottom: 1em;"],
    ["ml", "margin-left", "1em", "margin-left: 1em;"],
    ["mr", "margin-right", "1em", "margin-right: 1em;"],
    ["mt", "margin-top", "1em", "margin-top: 1em;"],
    ["mb", "margin-bottom", "1em", "margin-bottom: 1em;"],
  ])("should correctly handle prop %s", (prop, attr, value, expected) => {
    const props = {
      [prop]: value,
    };

    const actual = responsive(attr, (props: TestProps) => props[prop])(props);

    expect(stringify(actual)).toEqual(stringify(expected));
  });
});
