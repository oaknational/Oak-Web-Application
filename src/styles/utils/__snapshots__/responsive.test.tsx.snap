// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`responsive should return css for a single value 1`] = `
<div>
  <div
    class="sc-7b9f1971-0 eWCMdJ"
    data-testid="test"
  />
</div>
`;
